### PR TITLE
Wire PrintFlags through delete, replace, run commands

### PR DESCRIPTION
**Pull request by juanvallejo**
**Release note**:
```release-note
NONE
```

Wires PrintFlags through the `delete`, `replace`, and `run` commands.
All three commands grouped in this patch as they depend on DeleteOptions.

~~Tagged as WIP for now, as I still need to update tests.~~

cc @soltysh @deads2k

### CODE REVIEW

File: pkg/kubectl/cmd/cmd_test.go [hunk 1]
```diff
@@ -194,6 +194,7 @@ func Example_printMultiContainersReplicationControllerWithWide() {
 		},
 	}
 	cmd.Flags().Set("output", "wide")
+	cmdutil.AddNoHeadersFlags(cmd)
```
Comment (juanvallejo) [inline pkg/kubectl/cmd/cmd_test.go]: @soltysh fyi
This test expects human-readable output from the `run` command (output `wide` and `no-headers` flag).
Will update test to expect success message.

**Comment by juanvallejo** [comment]
@soltysh this should now be ready for review, ptal

File: hack/make-rules/test-cmd-util.sh [hunk 2]
```diff
@@ -3924,7 +3924,7 @@ run_cmd_with_img_tests() {
 
   # Test that a valid image reference value is provided as the value of --image in `kubectl run <name> --image`
   output_message=$(kubectl run test1 --image=validname)
-  kube::test::if_has_string "${output_message}" 'deployment.apps "test1" created'
+  kube::test::if_has_string "${output_message}" 'deployment.extensions/test1 created'
```
Comment (soltysh) [inline hack/make-rules/test-cmd-util.sh]: This is not acceptable, we cannot break backwards compatibility. Move the run changes to a separate PR and I'll have a closer look once again on Monday.

File: pkg/kubectl/cmd/cmd_test.go [hunk 3]
```diff
@@ -161,7 +161,7 @@ func Example_printMultiContainersReplicationControllerWithWide() {
 		NegotiatedSerializer: ns,
 		Client:               nil,
 	}
-	cmd := NewCmdRun(tf, os.Stdin, os.Stdout, os.Stderr)
+	cmd := resource.NewCmdGet(tf, os.Stdout, os.Stderr)
```
Comment (soltysh) [inline pkg/kubectl/cmd/cmd_test.go]: Drop all 4 tests, we don't guarantee output stability. Additionally, the server-side print is already tested as part of this test: https://github.com/kubernetes/kubernetes/blob/master/test/integration/apiserver/print_test.go

File: pkg/kubectl/cmd/delete.go [hunk 4]
```diff
@@ -233,17 +241,16 @@ func (o *DeleteOptions) Validate(cmd *cobra.Command) error {
 }
 
 func (o *DeleteOptions) RunDelete() error {
-	shortOutput := o.Output == "name"
 	// By default use a reaper to delete all related resources.
 	if o.Cascade {
-		return ReapResult(o.Result, o.f, o.Out, true, o.IgnoreNotFound, o.Timeout, o.GracePeriod, o.WaitForDeletion, shortOutput, false)
+		return o.ReapResult(o.Result, o.f, true, false)
```
Comment (soltysh) [inline pkg/kubectl/cmd/delete.go]: Add a TODO here, and explain why you're passing `o.Result` here.

Comment (soltysh) [reply pkg/kubectl/cmd/delete.go]: Get rid of factory! All this method cares about is reaper, so that's what you should put in `DeleteOptions` instead of full factory.

File: pkg/kubectl/cmd/replace.go [hunk 5]
```diff
@@ -71,61 +99,102 @@ func NewCmdReplace(f cmdutil.Factory, out io.Writer) *cobra.Command {
...
 	}
+	o.PrintObj = func(obj runtime.Object) error {
+		return printer.PrintObj(obj, o.Out)
+	}
 
-	force := cmdutil.GetFlagBool(cmd, "force")
-	if cmdutil.IsFilenameSliceEmpty(options.Filenames) {
-		return cmdutil.UsageErrorf(cmd, "Must specify --filename to replace")
+	// complete delete options
+	// TODO(juanvallejo): Turn these fields in a DeleteFlags struct, similar to PrintFlags
```
Comment (soltysh) [inline pkg/kubectl/cmd/replace.go]: Can you just do it, esp. that you are touching that.

File: pkg/kubectl/cmd/replace.go [hunk 6]
```diff
@@ -71,61 +99,102 @@ func NewCmdReplace(f cmdutil.Factory, out io.Writer) *cobra.Command {
 		Example: replaceExample,
 		Run: func(cmd *cobra.Command, args []string) {
 			cmdutil.CheckErr(cmdutil.ValidateOutputArgs(cmd))
-			err := RunReplace(f, out, cmd, args, options)
-			cmdutil.CheckErr(err)
+			cmdutil.CheckErr(options.Complete(f, cmd))
+			cmdutil.CheckErr(options.Validate(cmd))
+			cmdutil.CheckErr(options.Run(f, args))
```
Comment (soltysh) [inline pkg/kubectl/cmd/replace.go]: Don't pass factory here, extract whatever you need in `Complete` and use that in `Run`. The same applies to `args`.

File: pkg/kubectl/cmd/replace.go [hunk 7]
```diff
@@ -71,61 +99,102 @@ func NewCmdReplace(f cmdutil.Factory, out io.Writer) *cobra.Command {
...
 
+	if cmdutil.IsFilenameSliceEmpty(o.FileNameOptions.Filenames) {
+		return cmdutil.UsageErrorf(cmd, "Must specify --filename to replace")
+	}
+
+	return nil
+}
+
+func (o *ReplaceOpts) Run(f cmdutil.Factory, args []string) error {
+	schema, err := f.Validator(o.validate)
```
Comment (soltysh) [inline pkg/kubectl/cmd/replace.go]: This belongs in `Complete`.

File: pkg/kubectl/cmd/replace.go [hunk 8]
```diff
@@ -71,61 +99,102 @@ func NewCmdReplace(f cmdutil.Factory, out io.Writer) *cobra.Command {
...
+	if err != nil {
+		return err
+	}
+
+	cmdNamespace, enforceNamespace, err := f.DefaultNamespace()
```
Comment (soltysh) [inline pkg/kubectl/cmd/replace.go]: This belongs in `Complete`.

File: pkg/kubectl/cmd/replace.go [hunk 9]
```diff
@@ -71,61 +99,102 @@ func NewCmdReplace(f cmdutil.Factory, out io.Writer) *cobra.Command {
...
+	if err != nil {
+		return err
+	}
+
+	if o.DeleteOptions.ForceDeletion {
+		return o.forceReplace(f, args)
+	}
+
 	r := f.NewBuilder().
```
Comment (soltysh) [inline pkg/kubectl/cmd/replace.go]: This belongs in `Complete`.

File: pkg/kubectl/cmd/replace.go [hunk 10]
```diff
@@ -154,13 +223,12 @@ func RunReplace(f cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
...
-		cmdutil.PrintSuccess(shortOutput, out, info.Object, false, "replaced")
-		return nil
+		return o.PrintObj(info.AsVersioned())
 	})
 }
 
-func forceReplace(f cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string, shortOutput bool, options *resource.FilenameOptions) error {
-	schema, err := f.Validator(cmdutil.GetFlagBool(cmd, "validate"))
+func (o *ReplaceOpts) forceReplace(f cmdutil.Factory, args []string) error {
+	schema, err := f.Validator(o.validate)
```
Comment (soltysh) [inline pkg/kubectl/cmd/replace.go]: Why do you care about the `Validator` call if you already did that before getting here?

File: pkg/kubectl/cmd/run.go [hunk 11]
```diff
@@ -85,25 +86,62 @@ var (
 )
 
 type RunObject struct {
```
Comment (soltysh) [inline pkg/kubectl/cmd/run.go]: Like I said before split this part to a separate PR, I'm not even looking at it atm.

**Review by soltysh — CHANGES_REQUESTED** [review]

File: pkg/kubectl/cmd/replace.go (hunk 6 above)
Comment (juanvallejo) [reply pkg/kubectl/cmd/replace.go]: We need the factory here, as it's used to obtain a resource builder.
Args are also used by a resource builder in the `o.forceReplace` method.

File: pkg/kubectl/cmd/replace.go (hunk 10 above)
Comment (juanvallejo) [reply pkg/kubectl/cmd/replace.go]: We are obtaining a schema from it to use in the builder. I have moved this to a field in the options struct

File: pkg/kubectl/cmd/replace.go (hunk 6 above)
Comment (soltysh) [reply pkg/kubectl/cmd/replace.go]: Builder can be part of options, similarly to https://github.com/kubernetes/kubernetes/blob/ee4d90aaa61150139cdcd67a73e22da8cb226dc6/pkg/kubectl/cmd/create/create_job.go#L58

**Review by soltysh — APPROVED** [review]
/lgtm
